cpu/sam0_common/periph/usbdev.c

```diff
@@ -246,6 +246,45 @@ static inline void _poweron(sam0_common_usb_t *dev)
     GCLK->CLKCTRL.reg = GCLK_CLKCTRL_CLKEN
                       | GCLK_CLKCTRL_GEN(dev->config->gclk_src)
                       | GCLK_CLKCTRL_ID(USB_GCLK_ID);
+#elif defined(CPU_FAM_SAML21)
+    /* enable DFLL */
+    OSCCTRL->DFLLCTRL.reg = OSCCTRL_DFLLCTRL_ENABLE;
+    while (!(OSCCTRL->STATUS.reg & OSCCTRL_STATUS_DFLLRDY)) {}
+
+    /* setup initial values for closed loop */
+    OSCCTRL->DFLLVAL.reg = OSCCTRL_DFLLVAL_COARSE((*(uint32_t*)NVMCTRL_OTP5)
+                           >> 26) |  OSCCTRL_DFLLVAL_FINE(512);
+    while (!(OSCCTRL->STATUS.reg & OSCCTRL_STATUS_DFLLRDY)) {}
+
+    /* we will generate a 48 Mhz clock from 1KHz SOF signal */
+    OSCCTRL->DFLLMUL.reg = OSCCTRL_DFLLMUL_CSTEP(0x08) |
+                           OSCCTRL_DFLLMUL_FSTEP(0x08) |
+                           OSCCTRL_DFLLMUL_MUL((48000000U/1000));
+
+    /* Disable DFLL before setting its configuration */
+    OSCCTRL->DFLLCTRL.reg = 0;
+    while (!(OSCCTRL->STATUS.reg & OSCCTRL_STATUS_DFLLRDY)) {}
+
+    /* Configure DFLL for USB clock recovery mode */
+    OSCCTRL->DFLLCTRL.reg =  OSCCTRL_DFLLCTRL_USBCRM |
+                             OSCCTRL_DFLLCTRL_MODE |
+                             OSCCTRL_DFLLCTRL_CCDIS |
+                             OSCCTRL_DFLLCTRL_ENABLE;
+
+    /* Ensure COARSE and FINE are locked */
+    while((!(OSCCTRL->STATUS.bit.DFLLLCKC)) && (!(OSCCTRL->STATUS.bit.DFLLLCKF))) {}
+    while (!(OSCCTRL->STATUS.bit.DFLLRDY)) {}
+
+    /* set clock 1 to 48 MHz DFLL */
+    GCLK->GENCTRL[1].reg = GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_DFLL48M;
+    while (GCLK->SYNCBUSY.reg & GCLK_SYNCBUSY_GENCTRL(1)) {}
+
+    /* use it as USB clock */
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN_GCLK1;
+#elif defined(CPU_FAM_SAMD5X)
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN_GCLK6;
```

Comment:
This whole part should be rewrite with #10075 merged, the DFLL should already be used to source the CPU clock I think we should just connect the USB peripheral to GCLK0.
Further improvements could be made later.

Reply:
You probably want to introduce a separate `SAM0_GCLK_48MHZ` to which the DFLL is connected.
So you can still run the CPU at 12 MHz if you want.